common/transport/amqp/src/amqp.ts

```diff
@@ -132,6 +88,185 @@ export class Amqp {
         forever: false
       }
     }, amqp10.Policy.EventHub));
+
+    this._fsm = new machina.Fsm({
+      initialState: 'disconnected',
+      states: {
+        disconnected: {
+          connect: (connectCallback) => {
+            this._fsm.transition('connecting');
+            let connectError = null;
+            const connectErrorHandler = (err) => {
+              connectError = err;
+            };
+
+            const amqpErrorHandler = (err) => {
+              debug('amqp10 error: ' + err.toString());
+              if (this._disconnectHandler) {
+                this._disconnectHandler(err);
+              }
+            };
+
+            this._amqp.on(_amqpClientError, connectErrorHandler);
+            this._amqp.connect(this.uri)
+              .then((result) => {
+                debug('AMQP transport connected.');
+                this._amqp.on(_amqpClientError, amqpErrorHandler);
+                this._amqp.removeListener(_amqpClientError, connectErrorHandler);
+                this._fsm.transition('connected');
+                /*Codes_SRS_NODE_COMMON_AMQP_16_002: [The `connect` method shall establish a connection with the IoT hub instance and if given as argument call the `done` callback with a null error object in the case of success and a `results.Connected` object.]*/
+                this._safeCallback(connectCallback, null, new results.Connected(result));
+                return null;
+              })
+              .catch((err) => {
+                this._amqp.removeListener(_amqpClientError, connectErrorHandler);
+                this._fsm.transition('disconnected');
+                /*Codes_SRS_NODE_COMMON_AMQP_16_003: [The `connect` method shall call the `done` callback if the connection fails.] */
+                this._safeCallback(connectCallback, connectError || err);
+              });
+          },
+          disconnect: (callback) => callback(null, new results.Disconnected()),
+          attachSenderLink: (endpoint, linkOptions, callback) => {
+            this._fsm.handle('connect', (err) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('attachSenderLink', endpoint, linkOptions, callback);
+              }
+            });
```

Comment:
General comment.  I guess I would have done the above actual .connect call in the on handler of the connecting state.  I would have had all the appropriate handlers transition to the connecting and in connecting defer unit transition to connected.